src/bus/event_bus.hpp

```diff
@@ -0,0 +1,181 @@
+#ifndef SERENITY_EVENT_BUS_HPP
+#define SERENITY_EVENT_BUS_HPP
+
+#include <map>
+#include <unordered_set>
+// TODO(skonefal): Move to std shared_mutex when -std=c++14 is available
+#include <mutex>  // NOLINT [build/c++11]
+#include <string>
+#include <type_traits>
+#include <typeinfo>
+#include <typeindex>
+
+#include "glog/logging.h"
+
+#include "messages/serenity.hpp"
+
+#include "process/process.hpp"
+#include "process/protobuf.hpp"
+
+#include "serenity/serenity.hpp"
+
+#include "stout/error.hpp"
+#include "stout/nothing.hpp"
+#include "stout/option.hpp"
+
+namespace mesos {
+namespace serenity {
+
+
+/**
+ * Helper snippet of code for internal Envelope message type evaluation.
+ *
+ * Each event defined in proto need to be wrapped in "Envelope"
+ * with message variable e.g:
+ *
+ * message SomeEventEnvelope {
+ *  message SomeEvent {
+ *   optional bool first_event_param = 1;
+ *   optional bool second_event_param = 2;
+ *   (...)
+ *  }
+ *
+ *  optional SomeEvent message = 1;
+ * }
+ */
+template <class Envelope>
+using MessageType =
+  typename std::remove_const<
+    typename std::remove_reference<decltype(std::declval<Envelope>().message())
+    >::type
+  >::type;
+
+
+/**
+ *  Lookup Event-based bus made as libprocess actor.
+ *  Used as singleton.
+ */
+class EventBus : public ProtobufProcess<EventBus> {
+ public:
+  virtual ~EventBus() { }
+
+  /**
+   * Creates a new instance of the EventBus if hasn't already been created.
+   * Returns the singleton instance.
+   */
+  static EventBus* const GetInstance() {
+    if (EventBus::instance == nullptr) {
```

Comment:
To answer my own question: Go get double-checked locking to make it safe:
```
if (EventBus::instance == nullptr) {
  lock_guard<mutex> lock(subscribersMapLock);
  if (EventBus::instance == nullptr) {
    EventBus::instance = new EventBus();
    process::spawn(EventBus::instance);
  }
}
```
Also: you can do std::call_once, but DCL should be more convenient.